website/blog/2022-05-10-faster-presto-queries-with-parquet-page-index.md

```diff
@@ -0,0 +1,54 @@
+
+![Effectiveness Comparison: Column Statistics accept, Page Statistics partially accept](/img/blog/2022-05-10-faster-presto-queries-with-parquet-page-index/image5.png)
+
+In the third example, the data and statistics in both column chunk and page are still the same as the first example, but the filter is changed to ‘x =  700’. The column chunk statistic still determines ‘yes’ because it falls into its range [-100, 1000], but all the page statistics determine ‘no’ because none of them have a range that can cover the value ‘700’ that we are looking for. As a result, the whole column data is skipped for reading. 
+
+![Effectiveness Comparison: Column Statistics accept, Page Statistics reject](/img/blog/2022-05-10-faster-presto-queries-with-parquet-page-index/image6.png)
+
+In the above examples, we only show 3 pages in a column chunk, but in reality generally it has tens or hundreds pages in a column chunk. So the savings would be larger in reality. If the column data is sorted, the page statistics work the better because the possibility of false alarm is reduced.   
+
+## Page Statistics in Presto
+The page statistics in older versions than 1.11.0 are placed in the page header. The problem is that the reader has to read each individual page to get the page statistics. Then even it determines later not to read the page, but it is already read. The parquet-mr 1.11.0 fixed this issue by placing all the page statistics for a column chunk into one place, so that the reader can read them at once and make determination which page should be read.
+Because Presto partially rewrites the parquet-mr code, we need to port the changes in Parquet 1.11.0 into the Presto code base. The code change(PR-17284) has been merged into Presto master branch in Feb 2022 and will be released in the 0.273 version.
+
+## Benchmark Result
+We benchmarked the Parquet column index based on one of our heavily accessed production tables. We created a test table that is a snapshot of the original production tables with some partitions, and the test table is sorted. We then run Presto queries on the table from our staging environment. The Page Index feature can be opted on and off on a per query level using Presto session property. This allows us to do side-to-side comparison.
+
+We observed great improvement potential on input data scan. The figures below show the stats of running one example query when Parquet Paeg Index is enabled (left) vs disabled (right). As can be seen, they generated the same data for this stage, but it had to scan 149.31MB/239K rows raw input data when Parquet column index was not enabled, while it only needed to scan 63.31MB/75.4K rows raw input data if we enable Parquet column index. This is a 57% input read volume reduction. This is because when using Parquet Page Index, the operator can identify and skip pages that did not fall into the filter condition.
```

Comment:
```suggestion
We observed great improvement potential on input data scan. The figures below show the stats of running one example query when Parquet Page Index is enabled (left) vs disabled (right). As can be seen, they generated the same data for this stage, but it had to scan 149.31MB/239K rows raw input data when Parquet column index was not enabled, while it only needed to scan 63.31MB/75.4K rows raw input data if we enable Parquet column index. This is a 57% input read volume reduction. This is because when using Parquet Page Index, the operator can identify and skip pages that did not fall into the filter condition.
```
Quantities could be code or bold or something to just format them a little nicer from regular text.